### PR TITLE
Fix units encoding error

### PR DESCRIPTION
**Pull request by Tom-Willemsen**
Closes https://github.com/scipp/scippneutron/issues/140

This PR ensures that we always load string datasets and attributes with the encodings specified in the nexus file.

For ASCII-encoded data in the nexus file, it allows the common extended ascii (latin-1) format, but warns as this is technically unspecified behaviour. Data files from ISIS do sometimes use these extended ASCII code points (see example LET data file linked to ticket). Data marked as UTF-8 will still be loaded correctly using that encoding.

See also https://docs.h5py.org/en/stable/strings.html

### CODE REVIEW

File: python/src/scippneutron/file_loading/_hdf5_nexus.py [hunk 1]
```diff
@@ -10,19 +11,55 @@
...
+
+    if encoding == "ascii":
+        try:
+            return str(str_or_bytes, encoding="ascii")
+        except UnicodeDecodeError as e:
+            decoded = str(str_or_bytes, encoding="latin-1")
+            warnings.warn(f"Encoding for bytes '{str_or_bytes}' declared as ascii, "
+                          f"but contains characters in extended ascii range. Assuming "
+                          f"extended ASCII (latin-1), but this behaviour is not "
+                          f"specified by the HDF5 or nexus standards and may therefore "
```
Comment (OwenArnold) [inline python/src/scippneutron/file_loading/_hdf5_nexus.py:57]: Minor point, we are using American English spelling on the project `behaviour` -> `behavior`

File: python/src/scippneutron/file_loading/_hdf5_nexus.py [hunk 2]
```diff
@@ -10,19 +11,55 @@
 from ._common import Group, MissingDataset, MissingAttribute
 
 
+def cset_to_encoding(cset: int):
```
Comment (OwenArnold) [inline python/src/scippneutron/file_loading/_hdf5_nexus.py:14]: minor rename `_cset_to_encoding` also you are inconsistently applying type hinting for the return types in these functions, though you do for most arguments.

File: python/tests/nexus_helpers.py [hunk 3]
```diff
@@ -36,25 +36,25 @@ class EventData:
     event_time_offset: Optional[np.ndarray]
     event_time_zero: Optional[np.ndarray]
     event_index: Optional[np.ndarray]
-    event_time_zero_unit: Optional[str] = "ns"
-    event_time_zero_offset: Optional[str] = "1970-01-01T00:00:00Z"
-    event_time_offset_unit: Optional[str] = "ns"
+    event_time_zero_unit: Optional[Union[str, bytes]] = "ns"
```
Comment (OwenArnold) [inline python/tests/nexus_helpers.py:39]: Minor: Might be worth defining this Union somewhere common as is heavily used - though maybe doesn't make it much shorter or clearer?

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Overall I think what you have done makes sense.

File: python/tests/nexus_helpers.py (hunk 3 above)
Comment (SimonHeybrock) [reply python/tests/nexus_helpers.py:39]: I'd rather have this explicit than having to look up another name.

Comment (Tom-Willemsen) [reply python/tests/nexus_helpers.py:39]: I think on balance I agree with @SimonHeybrock that it's easier to understand as is? I can see the arguments both ways though and happy to change if you feel strongly about it.

**Comment by SimonHeybrock** [comment]
@OwenArnold Did you have more comments on this?

File: python/src/scippneutron/file_loading/_hdf5_nexus.py [hunk 4]
```diff
@@ -10,19 +11,66 @@
...
+    """
+    if cset == h5py.h5t.CSET_ASCII:
+        return "ascii"
+    elif cset == h5py.h5t.CSET_UTF8:
+        return "utf-8"
+    else:
+        warnings.warn(f"Unknown character set in HDF5 data file. Expected data "
+                      f"types are {h5py.h5t.CSET_ASCII} (H5T_CSET_ASCII) or "
+                      f"{h5py.h5t.CSET_UTF8} (H5T_CSET_UTF8) but got '{cset}'. "
+                      f"Assuming data is UTF-8 encoded but this may be incorrect.")
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply python/src/scippneutron/file_loading/_hdf5_nexus.py]: No, not aware of any facilities that do this. It's not actually allowed by the nexus standard. I've changed it to an error.

**Pull request merged**